Support/check Maven 3.9.0

~~Not out yet, but probably soon.~~

Changes to check/revisit:
- https://issues.apache.org/jira/browse/MNG-7068
- https://issues.apache.org/jira/browse/MNG-6972
- https://issues.apache.org/jira/browse/MNG-6965

----
> ### :warning: Action required to run GIB with Maven 3.9.0!
> The changes of https://issues.apache.org/jira/browse/MNG-6965 lead to compatibiliy issues across different Maven versions.
> Details can be found here: https://lists.apache.org/thread/wcbz8nsrrrdx8s8byoqpj99ksv73scqy
> 
> If you are using GIB 4.3.0 [via `extension.xml`](https://maven.apache.org/ref/3.8.7/maven-embedder/core-extensions.html) then you should be fine!
> 
> But if you are using GIB 4.3.0 via `pom.xml` (as `plugin` or `extension`), then you are probably seeing this error with Maven 3.9.0:
> ```
> [ERROR] Internal error: com.google.inject.ProvisionException: Unable to provision, see the following errors:
> [ERROR] 
> [ERROR] 1) [Guice/NullInjectedIntoNonNullable]: null returned by binding at LocatorWiring
> [ERROR]  but UnchangedProjectsRemover.downstreamCalculator(UnchangedProjectsRemover.java:43) is not @Nullable
> [ERROR]   at LocatorWiring
> [ERROR]   at UnchangedProjectsRemover.downstreamCalculator(UnchangedProjectsRemover.java:43)
> [ERROR]       \_ for field downstreamCalculator
> [ERROR]   at ClassRealm[extension>io.github.gitflow-incremental-builder:gitflow-incremental-builder:4.3.1-SNAPSHOT, parent: ClassLoaders$AppClassLoader@5cb0d902]
> [ERROR]       \_ installed by: WireModule -> PlexusBindingModule
> [ERROR]   at MavenLifecycleParticipant.unchangedProjectsRemover(MavenLifecycleParticipant.java:21)
> [ERROR]       \_ for field unchangedProjectsRemover
> [ERROR]   at ClassRealm[extension>io.github.gitflow-incremental-builder:gitflow-incremental-builder:4.3.1-SNAPSHOT, parent: ClassLoaders$AppClassLoader@5cb0d902]
> [ERROR]       \_ installed by: WireModule -> PlexusBindingModule
> [ERROR]   while locating MavenLifecycleParticipant
> [ERROR]   while locating Object annotated with *
> ```
> You can **avoid this error by adding the `plexus-utils` dependency** (*but only for the `plugin` usage case* as you cannot add a `dependency` to an `extension`):
> ```xml
> <build>
>     <plugins>
>         <plugin>
>             <groupId>io.github.gitflow-incremental-builder</groupId>
>             <artifactId>gitflow-incremental-builder</artifactId>
>             <version>4.3.0</version>
>             <extensions>true</extensions>
>             <configuration>
>                 <!-- ... -->
>             </configuration>
>             <dependencies>
>                 <dependency>
>                     <groupId>org.codehaus.plexus</groupId>
>                     <artifactId>plexus-utils</artifactId>
>                     <version>3.5.0</version>
>                 </dependency>
>             </dependencies>
>         </plugin>
>         <!-- ... -->
>     </plugins>
>     <!-- ... -->
> </build>
> ```
> GIB 4.*4*.0 will probably declare that dependency on its own, which will unfortunately break the `extension.xml` usage case with Maven 3.8.7 or below. Maven 3.8.8 might contain a fix for that (yet to be tested).

--- a/.github/workflows/maven.yml
+++ b/.github/workflows/maven.yml
@@ -113,6 +113,11 @@ jobs:
             maven-version: 3.8.7
           }
           - {
+            name: "JDK 17 & Maven 3.9.0",
+            java-version: 17,
+            maven-version: 3.9.0
+          }
+          - {
             name: "JDK 17 & Maven 3.6.3",
             java-version: 17,
             maven-version: 3.6.3

--- a/README.md
+++ b/README.md
@@ -72,7 +72,7 @@ This extension is **not limited to Git Flow setups!** The [extensive configurati
 
 To be able to use GIB, your project must use:
 
-- **Apache Maven** build tool, version **3.8.7** is recommended
+- **Apache Maven** build tool, version **3.8.7** or 3.9.0 is recommended
   - The minimum Maven version is 3.6.3 due to [MNG-6580](https://issues.apache.org/jira/browse/MNG-6580)
 
 - **Git** version control

--- a/pom.xml
+++ b/pom.xml
@@ -603,6 +603,8 @@
                 <configuration>
                     <excludes>
                         <exclude>**/mojo/**</exclude>
+                        <!-- unit tests will always use the official graph, but our graph is only a copy anyway -->
+                        <exclude>**/Maven38DefaultDependencyGraph.class</exclude>
                     </excludes>
                 </configuration>
                 <executions>
@@ -634,17 +636,18 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.91</minimum>
+                                            <minimum>0.93</minimum>
                                         </limit>
                                         <limit>
                                             <counter>BRANCH</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.84</minimum>
+                                            <minimum>0.85</minimum>
                                         </limit>
                                         <limit>
                                             <counter>CLASS</counter>
                                             <value>MISSEDCOUNT</value>
-                                            <maximum>0</maximum>
+                                            <!-- Maven38DefaultDependencyGraph is not covered, exclude does not seem to help here -->
+                                            <maximum>1</maximum>
                                         </limit>
                                     </limits>
                                 </rule>

--- a/pom.xml
+++ b/pom.xml
@@ -3,7 +3,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>io.github.gitflow-incremental-builder</groupId>
     <artifactId>gitflow-incremental-builder</artifactId>
-    <version>4.3.1-SNAPSHOT</version>
+    <version>4.4.0-SNAPSHOT</version>
     <name>${project.groupId}:${project.artifactId}</name>
     <description>A Maven extension for incremental building of multi-module projects when using Git Flow (or Git in general).</description>
     <url>https://github.com/gitflow-incremental-builder/gitflow-incremental-builder</url>
@@ -66,6 +66,7 @@
         <version.slf4j>1.7.36</version.slf4j>   <!-- must match the version that is provided by maven -->
         <version.logback>1.2.11</version.logback>    <!-- the latest that appears to work properly with above slf4j version -->
         <version.javax.inject>1</version.javax.inject>
+        <version.plexus-utils>3.5.1</version.plexus-utils>
         <version.bytebuddy>1.14.0</version.bytebuddy>
         <version.junit>5.9.2</version.junit>
         <version.archunit>1.0.1</version.archunit>
@@ -166,6 +167,11 @@
             <artifactId>javax.inject</artifactId>
             <version>${version.javax.inject}</version>
             <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.codehaus.plexus</groupId>
+            <artifactId>plexus-utils</artifactId>
+            <version>${version.plexus-utils}</version>
         </dependency>
         <dependency>
             <groupId>net.bytebuddy</groupId>
@@ -628,7 +634,7 @@
                                         <limit>
                                             <counter>INSTRUCTION</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.92</minimum>
+                                            <minimum>0.91</minimum>
                                         </limit>
                                         <limit>
                                             <counter>BRANCH</counter>
@@ -710,23 +716,6 @@
                         </goals>
                         <configuration>
                             <expression>project.version</expression>    <!-- just something, does not matter -->
-                        </configuration>
-                    </execution>
-                </executions>
-            </plugin>
-            <!-- ancient plexus-utils:1.1 is "required" in IT repo because it's no declared: https://issues.apache.org/jira/browse/MNG-6965
-                 and it's not declared because of: https://issues.apache.org/jira/browse/MNG-7160 -->
-            <plugin>
-                <artifactId>maven-dependency-plugin</artifactId>
-                <executions>
-                    <execution>
-                        <id>ensure-plexus-utils-downloaded</id>
-                        <phase>integration-test</phase>
-                        <goals>
-                            <goal>get</goal>
-                        </goals>
-                        <configuration>
-                            <artifact>org.codehaus.plexus:plexus-utils:1.1</artifact>
                         </configuration>
                     </execution>
                 </executions>

--- a/src/main/java/io/github/gitflowincrementalbuilder/DownstreamCalculator.java
+++ b/src/main/java/io/github/gitflowincrementalbuilder/DownstreamCalculator.java
@@ -19,6 +19,7 @@ import javax.inject.Singleton;
 import org.apache.maven.artifact.Artifact;
 import org.apache.maven.execution.MavenSession;
 import org.apache.maven.execution.ProjectDependencyGraph;
+import org.apache.maven.graph.DefaultProjectDependencyGraph;
 import org.apache.maven.model.Dependency;
 import org.apache.maven.model.Exclusion;
 import org.apache.maven.model.Model;
@@ -53,13 +54,18 @@ class DownstreamCalculator {
 
     public Stream<MavenProject> streamProjectWithDownstreamProjects(MavenProject project, Configuration config) {
         if (graph == null) {
-            if (config.mavenSession.getProjects().size() != config.mavenSession.getAllProjects().size()) {
+            var allProjects = config.mavenSession.getAllProjects();
+            if (config.mavenSession.getProjects().size() != allProjects.size()) {
                 // The reactor has been trimmed/filtered, most likely because of -pl and so the default ProjectDependencyGraph
                 // will not give us any downstream dependencies for modules that are not part of the trimmed reactor.
                 // Therefore we need to create a separate graph that contains all modules.
-                // Note: Due to https://issues.apache.org/jira/browse/MNG-6972 we cannot use DefaultDependencyGraph directly.
                 try {
-                    graph = new Maven38DefaultDependencyGraph(config.mavenSession.getAllProjects());
+                    try {
+                        graph = new DefaultProjectDependencyGraph(allProjects);
+                    } catch (NoClassDefFoundError err) {
+                        // cannot use DPDG in maven < 3.9.0 (https://issues.apache.org/jira/browse/MNG-6972) so use our own copy
+                        graph = new Maven38DefaultDependencyGraph(allProjects);
+                    }
                 } catch (CycleDetectedException | DuplicateProjectException e) {
                     throw new IllegalStateException(e); // extremely unlikely
                 }

--- a/src/main/java/io/github/gitflowincrementalbuilder/DownstreamCalculator.java
+++ b/src/main/java/io/github/gitflowincrementalbuilder/DownstreamCalculator.java
@@ -1,5 +1,6 @@
 package io.github.gitflowincrementalbuilder;
 
+import java.lang.reflect.Method;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
@@ -190,11 +191,11 @@ class DownstreamCalculator {
         return testJarClassifiersCache.computeIfAbsent(String.valueOf(project.hashCode()), ignored -> project.getBuildPlugins().stream()
                 .filter(p -> "maven-jar-plugin".equals(p.getArtifactId()))
                 .flatMap(p -> p.getExecutions().stream().filter(e -> e.getGoals().contains(TEST_JAR)))
-                .map(e -> (Xpp3Dom) e.getConfiguration())
+                .map(e -> Xpp3DomWrapper.build(e.getConfiguration(), logger))
                 .map(d -> Optional.ofNullable(d)
-		                .map(xpp3Dom -> xpp3Dom.getChild("classifier"))
-		                .map(Xpp3Dom::getValue)
-		                .orElse(TEST_JAR_DEFAULT_CLASSIFIER))
+                        .map(xpp3Dom -> xpp3Dom.getChild("classifier"))
+                        .map(Xpp3DomWrapper::getValue)
+                        .orElse(TEST_JAR_DEFAULT_CLASSIFIER))
                 .collect(Collectors.toUnmodifiableSet()));
     }
 
@@ -265,6 +266,91 @@ class DownstreamCalculator {
             } catch (ExpressionEvaluationException e) {
                 LOGGER.warn("Failed to evaluate: " + expression, e);
                 return expression;
+            }
+        }
+    }
+
+    // wraps org.codehaus.plexus.util.xml.Xpp3Dom for either direct access or reflective access
+    static interface Xpp3DomWrapper {
+
+        Xpp3DomWrapper getChild(String name);
+        String getValue();
+
+        static Xpp3DomWrapper build(Object xpp3Dom, Logger logger) {
+            if (xpp3Dom == null) {
+                return null;
+            }
+            try {
+                return new Direct((Xpp3Dom) xpp3Dom);
+            } catch (ClassCastException e) {
+                logger.info("Trying to access org.codehaus.plexus.util.xml.Xpp3Dom via reflection to work around an issue related to MNG-6965 "
+                        + "and Maven 3.8.7 (or earlier); https://lists.apache.org/thread/wcbz8nsrrrdx8s8byoqpj99ksv73scqy\n"
+                        + "Consider upgrading to Maven 3.8.8 or higher, if available.");
+                logger.debug("Full exception:", e);
+                return new Reflective(xpp3Dom, logger);
+            }
+        }
+
+        static class Direct implements Xpp3DomWrapper {
+
+            private final Xpp3Dom dom;
+
+            Direct(Xpp3Dom dom) {
+                this.dom = dom;
+            }
+
+            @Override
+            public Xpp3DomWrapper getChild(String name) {
+                var child = dom.getChild(name);
+                return child != null ? new Direct(child) : null;
+            }
+
+            @Override
+            public String getValue() {
+                return dom.getValue();
+            }
+        }
+
+        static class Reflective implements Xpp3DomWrapper {
+
+            private static Method getChild;
+            private static Method getValue;
+            private final Object dom;
+            private final Logger logger;
+
+            Reflective(Object dom, Logger logger) {
+                this.dom = dom;
+                this.logger = logger;
+            }
+
+            @Override
+            public Xpp3DomWrapper getChild(String name) {
+                try {
+                    if (getChild == null) {
+                        getChild = dom.getClass().getMethod("getChild", String.class);
+                    }
+                    var child = getChild.invoke(dom, name);
+                    return child != null ? new Reflective(child, logger) : null;
+                } catch (ReflectiveOperationException | SecurityException e) {
+                    logger.warn("Unable to apply reflection workaround for Xpp3Dom: {}", e.toString());
+                    logger.debug("Full exception:", e);
+                    return null;
+                }
+            }
+
+            @Override
+            public String getValue() {
+                try {
+                    if (getValue == null) {
+                        getValue = dom.getClass().getMethod("getValue");
+                    }
+                    var value = getValue.invoke(dom);
+                    return value != null ? value.toString() : null;
+                } catch (ReflectiveOperationException | SecurityException e) {
+                    logger.warn("Unable to apply reflection workaround for Xpp3Dom: {}", e.toString());
+                    logger.debug("Full exception:", e);
+                    return null;
+                }
             }
         }
     }

--- a/src/main/java/io/github/gitflowincrementalbuilder/MavenLifecycleParticipant.java
+++ b/src/main/java/io/github/gitflowincrementalbuilder/MavenLifecycleParticipant.java
@@ -116,7 +116,7 @@ public class MavenLifecycleParticipant extends AbstractMavenLifecycleParticipant
     void warnIfBuggyOrUnsupportedMavenVersion(String mavenVersion, Configuration config) {
         if (mavenVersion == null) {
             logger.warn("Could not get Maven version.");
-        } else if (!mavenVersion.startsWith("3.8.") && !mavenVersion.equals("3.6.3")) {
+        } else if (!mavenVersion.startsWith("3.8.") && !mavenVersion.startsWith("3.9") && !mavenVersion.equals("3.6.3")) {
             logger.warn("Detected Maven {} was not tested with gitflow-incremental-builder.", mavenVersion);
         }
     }

--- a/src/test/java/io/github/gitflowincrementalbuilder/ArchitectureTest.java
+++ b/src/test/java/io/github/gitflowincrementalbuilder/ArchitectureTest.java
@@ -6,6 +6,7 @@ import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;
 import com.tngtech.archunit.base.DescribedPredicate;
 import com.tngtech.archunit.core.domain.JavaAccess;
 import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaModifier;
 import com.tngtech.archunit.core.importer.ImportOption.DoNotIncludeTests;
 import com.tngtech.archunit.junit.AnalyzeClasses;
 import com.tngtech.archunit.junit.ArchTest;
@@ -30,7 +31,10 @@ public class ArchitectureTest {
                     .and(new DescribedPredicate<JavaClass>("are only called from classes that reside in same package") {
                             @Override
                             public boolean test(final JavaClass cls) {
-                                return cls.getAccessesToSelf()
+                                var enclosingClassPublicOrNotPresent = cls.getEnclosingClass()
+                                        .map(eClass -> eClass.getModifiers().contains(JavaModifier.PUBLIC))
+                                        .orElse(true);
+                                return enclosingClassPublicOrNotPresent && cls.getAccessesToSelf()
                                     .stream()
                                     .map(JavaAccess::getOriginOwner)
                                     .allMatch(callerClass -> callerClass.getPackageName().equals(cls.getPackageName()));

--- a/src/test/java/io/github/gitflowincrementalbuilder/Xpp3DomWrapperTest.java
+++ b/src/test/java/io/github/gitflowincrementalbuilder/Xpp3DomWrapperTest.java
@@ -1,0 +1,44 @@
+package io.github.gitflowincrementalbuilder;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.junit.jupiter.api.Test;
+
+public class Xpp3DomWrapperTest {
+
+    @Test
+    public void reflective_noChild() {
+        var dom = new Xpp3Dom("configuration");
+
+        var underTest = new DownstreamCalculator.Xpp3DomWrapper.Reflective(dom, null);
+
+        assertThat(underTest.getChild("foo")).isNull();
+    }
+
+    @Test
+    public void reflective_noValue() {
+        var dom = new Xpp3Dom("configuration");
+        dom.addChild(new Xpp3Dom("foo"));
+
+        var underTest = new DownstreamCalculator.Xpp3DomWrapper.Reflective(dom, null);
+
+        var child = underTest.getChild("foo");
+        assertThat(child).isNotNull();
+        assertThat(child.getValue()).isNull();
+    }
+
+    @Test
+    public void reflective_hasValue() {
+        var dom = new Xpp3Dom("configuration");
+        var child = new Xpp3Dom("foo");
+        child.setValue("expected");
+        dom.addChild(child);
+
+        var underTest = new DownstreamCalculator.Xpp3DomWrapper.Reflective(dom, null);
+
+        var actualChild = underTest.getChild("foo");
+        assertThat(actualChild).isNotNull();
+        assertThat(actualChild.getValue()).isEqualTo(child.getValue());
+    }
+}

--- a/src/test/java/io/github/gitflowincrementalbuilder/integration/MavenIntegrationTestBase.java
+++ b/src/test/java/io/github/gitflowincrementalbuilder/integration/MavenIntegrationTestBase.java
@@ -374,7 +374,7 @@ public abstract class MavenIntegrationTestBase extends BaseRepoTest {
                 .doesNotContain("Building child1")
                 .doesNotContain("Building child4")
                 .contains("Building subchild41")
-                .contains("maven-compiler-plugin");
+                .contains("default-compile");
     }
 
     @Test

